enforcer/datapath/datapath_tcp.go

```diff
@@ -754,54 +716,57 @@ func (d *Datapath) createTCPAuthenticationOption(token []byte) []byte {
 
 // appSynRetrieveState retrieves state for the the application Syn packet.
 // It creates a new connection by default
-func (d *Datapath) appSynRetrieveState(p *packet.Packet) (*pucontext.PUContext, *connection.TCPConnection, error) {
+func (d *Datapath) appSynRetrieveState(p *packet.Packet) (*connection.TCPConnection, error) {
 
 	context, err := d.contextFromIP(true, p.SourceAddress.String(), p.Mark, strconv.Itoa(int(p.SourcePort)))
 	if err != nil {
-		return nil, nil, fmt.Errorf("no context in app processing: %s", err)
+		return nil, fmt.Errorf("No Context in App Processing")
 	}
 
 	conn, err := d.appOrigConnectionTracker.GetReset(p.L4FlowHash(), 0)
 	if err != nil {
-		conn = connection.NewTCPConnection()
-
+		conn = connection.NewTCPConnection(context)
 	}
 
-	conn.(*connection.TCPConnection).Lock()
-	conn.(*connection.TCPConnection).Context = context
-	conn.(*connection.TCPConnection).Unlock()
-	return context, conn.(*connection.TCPConnection), nil
+	return conn.(*connection.TCPConnection), nil
```

Comment:
You can avoid casting in the case you’ve just created a NewTCPConnection. Casting is not cheap

Reply:
Done